staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/integrator.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/apiserver/pkg/util/clock"
+)
+
+// An integrator computes the integral of some variable X over time.
+// The integral starts when the Integrator is created, and ends
+// at the latest call to Set or Add.
+type Integrator interface {
+	Set(float64) // set the value of X
+	Add(float64) // add the given quantity to X
+	GetResults() IntegratorResults
+}
+
+type IntegratorResults struct {
+	duration  float64 //seconds
+	average   float64
+	deviation float64 //sqrt(avg((value-avg)^2))
+}
+
+type integrator struct {
+	sync.Mutex
+	clk       clock.PassiveClock
+	lastTime  time.Time
+	x         float64
+	integrals [3]float64 // integral of x^0, x^1, and x^2
+}
+
+func NewIntegrator(clk clock.PassiveClock) Integrator {
+	return &integrator{
+		clk:      clk,
+		lastTime: clk.Now(),
+	}
+}
+
+func (igr *integrator) Set(x float64) {
```

Comment:
Not sure if it is intended but I think Set and Add are the same currently

Reply:
i dont think so? they're not the same

```
- igr.x = x
+ igr.x += deltaX
```

Reply:
No, they differ slightly.